src/Build/Instance/ProjectItemInstance.cs

```diff
@@ -1575,23 +1541,42 @@ public bool Equals(TaskItem other)
                     return true;
                 }
 
-                // Since both sides are this class, we know both sides support ITaskItem2.
-                ITaskItem2 thisAsITaskItem2 = this as ITaskItem2;
-                ITaskItem2 otherAsITaskItem2 = other as ITaskItem2;
+                ITaskItem2 thisAsITaskItem2 = this;
+                ITaskItem2 otherAsITaskItem2 = other;
 
                 // This is case-insensitive. See GetHashCode().
                 if (!MSBuildNameIgnoreCaseComparer.Default.Equals(thisAsITaskItem2.EvaluatedIncludeEscaped, otherAsITaskItem2.EvaluatedIncludeEscaped))
                 {
                     return false;
                 }
 
-                if (this.CustomMetadataCount != other.CustomMetadataCount)
+                var thisNames = new HashSet<string>(MSBuildNameIgnoreCaseComparer.Default);
+
+                if (_itemDefinitions is not null)
                 {
-                    return false;
+                    foreach (ProjectItemDefinitionInstance itemDefinition in _itemDefinitions)
+                    {
+                        thisNames.UnionWith(itemDefinition.MetadataNames);
+                    }
+                }
+
+                if (_directMetadata is not null)
+                {
+                    foreach (ProjectMetadataInstance metadatum in _directMetadata)
```

Comment:
Maybe add a comment that if this were values and not just names, you should be overwriting?

The only case I can think of when that would be relevant is if we had a bug where we used the case-sensitive string as a key into some dictionary instead of the case-insensitive string, in which case we might not find the right metadatum. (To be clear, I don't think we should be overly worried about that, but I think it may be worth a comment.)